core/src/main/java/com/datastax/oss/driver/internal/core/loadbalancing/BasicLoadBalancingPolicy.java

```diff
@@ -325,24 +338,63 @@ protected Queue<Node> maybeAddDcFailover(@Nullable Request request, @NonNull Que
 
           @Override
           protected Object[] computeNodes() {
-            Object[] remoteNodes =
-                liveNodes.dcs().stream()
-                    .filter(Predicates.not(Predicates.equalTo(localDc)))
-                    .flatMap(dc -> liveNodes.dc(dc).stream().limit(maxNodesPerRemoteDc))
-                    .toArray();
-
-            int remoteNodesLength = remoteNodes.length;
-            if (remoteNodesLength == 0) {
-              return EMPTY_NODES;
+            if (!preferredRemoteLocalDcs.isEmpty()) {
```

Comment:
Should the nodes within the same preferred remote DC be shuffled? For eg: If node1, node2 are in remote dc1 and node3, node4 in remote dc2 and maxNodesPerRemoteDc=2 and preference order specified is dc1, dc2.  In this example, the remote nodes in query plan will always be: node1, node2, node3, node4. Does it make sense to shuffle the nodes within a preferred remote DC, i.e it can be ((node2, node1), (node4, node3))?

Reply:
Yes, that is the big question here. I'm hesitating on the practical usefulness vs added complexity, but I think it makes sense to do as you suggest. If you take my code snippet above, this should be easily doable by changing the `flatMap` operator.

Reply:
However I realize I owe you an apology: using `Arrays.binarySearch` is wrong because `preferredRemoteDcs` is not necessarily sorted. We need `preferredRemoteDcs` to become a `List<String>` again and use `indexOf` instead:

```java
.sorted(
  Comparator.comparingInt(
      dc -> {
        int i = preferredRemoteDcs.indexOf(dc);
        return i < 0 ? Integer.MAX_VALUE : i;
      }))
```

Sorry for that 😅

Reply:
@adutra  Thank you for pointing it out. I have addressed the code review comments. Please take a look.